###################################################################################

I-TASSER-RNA is a program for accurate de novo RNA tertiary structure prediction

######################### INSTALLATION INSTRUCTIONS ############################
System requirements: x86_64 machine, Linux kernel OS, Free disk space of more 
	than 500GB. The vast majority of the disk space will be used to store the 
	sequence databases.
######################## General information on how to run I-TASSER-RNA #################
a, RNA 3D modeling needs 
   1. RNA sequence-structure local threading meta-alignment
      algorithm for full-length initial models and spatial constraints(xx)
   2. inter-nucleotide distance prediction by a deep residual network (xx)
   3. submit RNA3D simulation jobs for first-round simulation(xx)  
   4. cluster decoys and refine cluster centroid models for second-round simulation(xx)
   5. final model selection and all-atom model construction (xx)
 ######################## Each program in I-TASSER-RNA #################################
a, collect all templates by threading program and other predition programs
1. input features
   seq.fasta (query fasta)
2. output
   seq.afa (MSA from rMSA)
   seq.a2m_msa (MSA from RNAcamp3)
   seq.a2m (MSA from RNAcentral by Infernal)
   template/*pdb
   
b, inter-nucleotide geometires prediction
   1. input features 
      seq.fasta (query fasta)
      seq.afa (MSA from rMSA)
      seq.a2m_msa (MSA from RNAcamp3)
      seq.a2m (MSA from RNAcentral by Infernal)
   2. output
      npz_dist_p.txt  (distance prediction is best)
      npz_dist_c4.txt  (distance prediction is best)
      npz_dist_n.txt  (distance prediction is best)
      npz_oritention_theta.txt  (distance prediction is best)

c, folding simulation
   1. input files
      *_model.pdb (initial full-length  models built by threading templates)
      contactnn.dat (dNiNj<10 in two nucleotides with |i-j|>9nt;i j P N top40)
      contactpp.dat (dPiPj<20 in two nucleotides with |i-j|>9nt;i j P N top40)
      distL_8L.dat (i+j*8;i=1,2...L;J=2;top3)
      distL_10L.dat (i+j*10;top3)
      short_6L.dat (|i-j|<7;hard top40; easy top 10)
      save_280000_cb.txt (distance prediction)
   2. output 
      *conf.dat
      *U.dat
      *uN.dat	

d, optimize simulation
   1. input files
      combo*.dat (cluter centroids as initial models)
   2. output
      CG.dat

e, clustering
   1. input files
      trmsinp   (length of query)
      tseq.dat  (secondary structure information generated by SPOT_RNA)
      ttra.in   (prepare decoys for clustering)
   2. output
      combo.dat (cluster centeriod for each cluster)
      close1.dat (cluster center)

f, all-atom model construction
   1. input files
      frag_.dat (fragment of A,U,C,G by all atom)
      CG.dat  (coarse-grained model from simulation)
----------------------- Running QRNA step by step ---------------------------


2.  From the package directory, first install the third party programs by running the command:
	./scripts/install_third_party.sh

	Note, users are responsible for adhering to the licenses for these programs, 
		which include:
		
		PETfold: used for secondary structure prediction 
		(https://rth.dk/resources/petfold/)
		This program was developed Rolf Backofen's and SÃ¸ren Brun's Lab
		Citation: Seemann SE, Gorodkin J, Backofen R. Nucleic Acids Res., 
			36(20):6355-62, 2008

		rMSA: used for MSA generation (https://github.com/pylelab/rMSA)
		This program was developed by Chengxin Zhang at Anna Pyle's Lab	
                Citation: Zhang C, Zhang Y, Pyle AM (2021) rMSA: accurate multiple 
			sequence alignment generation to improve RNA structure modeling. 
			ISMB, webinar.

		
